debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java

```diff
@@ -0,0 +1,174 @@
+    public void execute(ChangeEventSourceContext context) throws InterruptedException {
+        if (!snapshotter.shouldStream()) {
+            LOGGER.info("Streaming is not enabled in currect configuration");
+            return;
+        }
+
+        try {
+            if (offsetContext.hasLastKnownPosition()) {
+                // start streaming from the last recorded position in the offset
+                final Long lsn = offsetContext.lsn();
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug("retrieved latest position from stored offset '{}'", ReplicationConnection.format(lsn));
+                }
+                replicationStream.compareAndSet(null, replicationConnection.startStreaming(lsn));
+            } else {
+                LOGGER.info("no previous LSN found in Kafka, streaming from the latest xlogpos or flushed LSN...");
+                replicationStream.compareAndSet(null, replicationConnection.startStreaming());
+            }
+
+            // refresh the schema so we have a latest view of the DB tables
+            taskContext.refreshSchema(connection, true);
+
+            this.lastCompletelyProcessedLsn = offsetContext.lsn();
+
+            final ReplicationStream stream = this.replicationStream.get();
+            while (context.isRunning()) {
+                if (!stream.readPending(message -> {
+                    final Long lsn = stream.lastReceivedLsn();
+                    if (message == null) {
+                        LOGGER.trace("Received empty message");
+                        lastCompletelyProcessedLsn = lsn;
+                        dispatcher.dispatchHeartbeatEvent(offsetContext);
+                        return;
+                    }
+                    if (message.isLastEventForLsn()) {
+                        lastCompletelyProcessedLsn = lsn;
+                    }
+
+                    final TableId tableId = PostgresSchema.parse(message.getTable());
+                    Objects.requireNonNull(tableId);
+
+                    offsetContext.updateWalPosition(lsn, lastCompletelyProcessedLsn, message.getCommitTime(), message.getTransactionId(), tableId, taskContext.getSlotXmin(connection));
+                    dispatcher
+                        .dispatchDataChangeEvent(
+                                tableId,
+                                new PostgresChangeRecordEmitter(
+                                        offsetContext,
+                                        clock,
+                                        connectorConfig,
+                                        schema,
+                                        connection,
+                                        message
+                                )
+                        );
+                })) {
+                    if (offsetContext.hasCompletelyProcessedPosition()) {
+                        dispatcher.dispatchHeartbeatEvent(offsetContext);
+                        pauseNoMessage.pause();
```

Comment:
@jpechane, should the `pause()` call be done in any case, also if the current LSN isn't fully processed yet? I'm wondering whether this creates some sort of "busy spinning" loop otherwise. WDYT?

Reply:
@gunnarmorling Right

Reply:
@gunnarmorling Fixed